docs/internals/optimizer.rst

```diff
@@ -65,20 +66,23 @@ A "runs" parameter of "1" will produce short but expensive code. The largest val
 Opcode-Based Optimizer Module
 =============================
 
-The opcode-based optimizer module operates on assembly. It splits the
+The opcode-based optimizer module operates on assembly code. It splits the
 sequence of instructions into basic blocks at ``JUMPs`` and ``JUMPDESTs``.
 Inside these blocks, the optimizer analyzes the instructions and records every modification to the stack,
 memory, or storage as an expression which consists of an instruction and
-a list of arguments which are pointers to other expressions. The opcode-based optimizer
-uses a component called "CommonSubexpressionEliminator" that amongst other
+a list of arguments which are pointers to other expressions. 
+
+Additionally, the opcode-based optimizer
+uses a component called "CommonSubexpressionEliminator" that, amongst other
 tasks, finds expressions that are always equal (on every input) and combines
 them into an expression class. It first tries to find each new
 expression in a list of already known expressions. If no such matches are found,
 it simplifies the expression according to rules like
 ``constant + constant = sum_of_constants`` or ``X * 1 = X``. Since this is
 a recursive process, we can also apply the latter rule if the second factor
-is a more complex expression where we know that it always evaluates to one.
-Modifications to storage and memory locations have to erase knowledge about
+is a more complex expression which we know always evaluates to one.
+
+Modifications to storage and memory locations must erase knowledge about
```

Comment:
I really struggled to understand this paragraph. I think it's saying something about a function that takes arguments x, and y, and does `SSTORE(X, VAL1)`, and  `SSTORE(Y, VAL2)`, but then I'm not sure what the optimizer does about that.

Reply:
The optimizer tries to symbolically track memory and storage. For example, this is how Keccak-256 may be evaluated during compile time, for example, consider the sequence:

```
 [PUSH 32, PUSH 0, CALLDATALOAD, PUSH 100,  DUP2, MSTORE, KECCAK256]
```

or the equivalent Yul
```
let x := calldataload(0)
mstore(x, 100)
let value := keccak256(x, 32)
```

In this case, the optimizer has to 'symbolically' track the value at a memory location `calldataload(0)` and then realize that the Keccak-256 has can be evaluated at compile time. This only works if there is no other instruction that modifies memory in between the `mstore` and `keccak256`. So if there is an instruction that writes to memory (or storage), then we need to erase the knowledge of the current memory (or storage). There is, however, an exception to this erasing, where we can easily see that the instruction doesn't write to a certain location.

For example,

```
let x := calldataload(0)
mstore(x, 100)
/// Current knowledge memory location x -> 100
let y := add(x, 32)
/// Does not clear the knowledge that x -> 100, since y does not write to [x, x + 32)
mstore(y, 200)
```

-----

I admit that the current paragraph is easy to understand if you already know how it works and rather confusing otherwise. Do you have a suggestion on rewording it?